### PR TITLE
Adjust yield/yield* typing to explicitly apply to each element

### PR DESCRIPTION
**Pull request by eernstg**
Cf. #829; if we decide to proceed as recommended there, this PR will do it.

### CODE REVIEW

File: specification/dartLangSpec.tex [hunk 1]
```diff
@@ -15646,17 +15648,24 @@ \subsubsection{Yield}
...
+Otherwise, if the body of $f$ is marked \code{\SYNC*} and
+the declared return type of $f$ implements \code{Iterable<$U$>} for some $U$
+then let $T_f$ be $U$.
+% $f$ is a generator function, so the return type must be a supertype of
+% `Iterable<U>`/`Stream<U>` for some `U`, so here it can only be a top type.
+Otherwise let $T_f$ be \DYNAMIC.
+It is a compile-time error if $T_e$ may not be assigned to $T_f$.
 
 
 \subsubsection{Yield-Each}
```
Comment (lrhn) [inline specification/dartLangSpec.tex]: Should we add an "it is a dynamic type error if the runtime type of $x$ is not a subtype of (the element type)" to the semantics of the yield*?

File: specification/dartLangSpec.tex [hunk 2]
```diff
@@ -15747,14 +15756,36 @@ \subsubsection{Yield-Each}
...
+\commentary{Let $T_f$ specify the type of elements that $f$ yields:}
+If the body of $f$ is marked \code{\ASYNC*} and
+the declared return type of $f$ implements \code{Stream<$U$>} for some $U$
+then let $T_f$ be $U$.
+Otherwise, if the body of $f$ is marked \code{\SYNC*} and
+the declared return type of $f$ implements \code{Iterable<$U$>} for some $U$
+then let $T_f$ be $U$.
+% $f$ is a generator function, so the return type must be a supertype of
+% `Iterable<U>`/`Stream<U>` for some `U`, so here it can only be a top type.
+Otherwise let $T_f$ be \DYNAMIC.
```
[the remark this answers is not in the document]
Comment (eernstg) [reply specification/dartLangSpec.tex]: Done. Also substantially cleaned up the text in section Yield and Yield-Each, similarly to other sections when they have been brought up to date (e.g., specifying the static analysis before dynamic semantics).

File: specification/dartLangSpec.tex (hunk 1 above)
Comment (eernstg) [reply specification/dartLangSpec.tex]: Done.

File: specification/dartLangSpec.tex [hunk 3]
```diff
@@ -15747,14 +15756,36 @@ \subsubsection{Yield-Each}
...
+\commentary{Let $T_e$ specify the type of elements that $e$ provides:}
+If the body of $f$ is marked \code{\ASYNC*} and
+$T$ implements \code{Stream<$U$>} for some $U$
+(\ref{interfaceSuperinterfaces})
+then let $T_e$ be $U$.
+Otherwise, if the body of $f$ is marked \code{\SYNC*} and
+$T$ implements \code{Iterable<$U$>} for some $U$
+then let $T_e$ be $U$.
+%% TODO(eernst): Come nnbd, change 'a top type' to \DYNAMIC{}.
+Otherwise if $T$ is a top type then let $T_e$ be \DYNAMIC.
```
[the remark this answers is not in the document]
Comment (eernstg) [reply specification/dartLangSpec.tex]: Done.

**Pull request merged**